django/db/models/aggregates.py

```diff
@@ -174,6 +175,29 @@ def __init__(self, expression, filter=None, **extra):
             raise ValueError("Star cannot be used with filter. Please specify a field.")
         super().__init__(expression, filter=filter, **extra)
 
+    def resolve_expression(self, *args, **kwargs):
+        result = super().resolve_expression(*args, **kwargs)
+        expr = result.source_expressions[0]
+
+        # In case of composite primary keys, count the first column.
+        if isinstance(expr, ColPairs):
+            if self.distinct:
+                raise NotSupportedError(
+                    "COUNT(DISTINCT) doesn't support composite primary keys"
+                )
+
+            cols = expr.get_cols()
+            fields = [col.field for col in cols]
+            pk_fields = expr.field.model._meta.pk_fields
+            assert fields == pk_fields, (
+                "Count doesn't support counting composite fields that are not composite"
+                "primary keys"
+            )
```

Comment:
We should raise a `NotSupportedError` instead of an `AssertionError`.

Reply:
Actually, as far as I understand, based on the discussion https://github.com/django/django/pull/18485#discussion_r1753906114 , we shouldn't do these type of checks in Django.